packages/datasets-fs/src/commands/generateV2.ts

```diff
@@ -0,0 +1,204 @@
+
+        const bar = new cliProgress.SingleBar({}, cliProgress.Presets.shades_classic)
+        bar.start(count, 0)
+
+        // generate n captchas
+        const captchas: CaptchaWithoutId[] = []
+        for (let i = 0; i < count; i++) {
+            bar.increment()
+            // this.logger.info(`generating captcha ${i + 1} of ${count}`)
+
+            if (this.targets.length <= 1) {
+                throw new ProsopoEnvError(
+                    new Error(`not enough different labels in labelled data`),
+                    'DATASET.NOT_ENOUGH_LABELS'
+                )
+            }
+
+            // uniformly sample targets
+            const target = at(this.targets, i % this.targets.length)
+            const notTargets = this.targets.filter((t) => t !== target)
+            // how many labelled images should be in the captcha?
+            const nLabelled = _.random(minLabelled, maxLabelled)
+            // how many correct labelled images should be in the captcha?
+            const maxCorrect = nLabelled - minCorrect
+            const nCorrect = _.random(minCorrect, maxCorrect)
+            const nIncorrect = nLabelled - nCorrect
+            const nUnlabelled = size - nLabelled
+
+            const targetItems = get(this.labelToImages, target)
+            const notTargetItems: Item[] = notTargets.map((notTarget) => get(this.labelToImages, notTarget)).flat()
+
+            if (nUnlabelled > this.unlabelled.length) {
+                throw new ProsopoEnvError(new Error(`not enough unlabelled data`), 'DATASET.NOT_ENOUGH_IMAGES')
+            }
+            if (nCorrect > targetItems.length) {
+                throw new ProsopoEnvError(
+                    new Error(`not enough images for target (${target})`),
+                    'DATASET.NOT_ENOUGH_IMAGES'
+                )
+            }
+            if (nIncorrect > notTargetItems.length) {
+                throw new ProsopoEnvError(
+                    new Error(`not enough non-matching images for target (${target})`),
+                    'DATASET.NOT_ENOUGH_IMAGES'
+                )
+            }
+
+            // get the correct items
+            const correctItems: Item[] = _.sampleSize(targetItems, nCorrect)
+
+            // get the incorrect items
+            const incorrectItems: Item[] = _.sampleSize(notTargetItems, nIncorrect)
+
+            // get the unlabelled items
+            const unlabelledItems = new Set<Item>()
+            while (unlabelledItems.size < size - nLabelled) {
+                // get a random image from the unlabelled data
+                const image = at(this.unlabelled, _.random(0, this.unlabelled.length - 1))
+                unlabelledItems.add(image)
+            }
+
+            let items: Item[] = [...correctItems, ...incorrectItems, ...unlabelledItems]
```

Comment:
Could we use functional methods here to avoid use of let

Reply:
done